File: backend/src/graph/graph.js
import { StateGraph, Annotation } from "@langchain/langgraph";

import {
  gptNode,
  deepseekNode,
  nvidiaNode,
  geminiNode,
  gptRefineNode,
  deepseekRefineNode,
  nvidiaRefineNode,
  geminiRefineNode,
  selectBestNode
} from "./nodes.js";

/* ================== STATE ================== */

const GraphState = Annotation.Root({
  question: Annotation(),

  gpt: Annotation(),
  deepseek: Annotation(),
  nvidia: Annotation(),
  gemini: Annotation(),

  gpt_refined: Annotation(),
  deepseek_refined: Annotation(),
  nvidia_refined: Annotation(),
  gemini_refined: Annotation(),

  gpt_confidence: Annotation(),
  deepseek_confidence: Annotation(),
  nvidia_confidence: Annotation(),
  gemini_confidence: Annotation(),

  finalAnswer: Annotation(),
  bestModel: Annotation(),
  refinedAnswers: Annotation()
});

/* ================== GRAPH ================== */

export function createGraph() {
  const graph = new StateGraph(GraphState);

  /* ===== BASE MODELS ===== */
  graph.addNode("runGPT", gptNode);
  graph.addNode("runDeepSeek", deepseekNode);
  graph.addNode("runNvidia", nvidiaNode);
  graph.addNode("runGemini", geminiNode);

  /* ===== REFINEMENT ===== */
  graph.addNode("refineGPT", gptRefineNode);
  graph.addNode("refineDeepSeek", deepseekRefineNode);
  graph.addNode("refineNvidia", nvidiaRefineNode);
  graph.addNode("refineGemini", geminiRefineNode);

 
  graph.addNode("selectBest", selectBestNode);

  graph.setEntryPoint("runGPT");

  graph.addEdge("runGPT", "runDeepSeek");
  graph.addEdge("runDeepSeek", "runNvidia");
  graph.addEdge("runNvidia", "runGemini");

  graph.addEdge("runGemini", "refineGPT");
  graph.addEdge("refineGPT", "refineDeepSeek");
  graph.addEdge("refineDeepSeek", "refineNvidia");
  graph.addEdge("refineNvidia", "refineGemini");

  graph.addEdge("refineGemini", "selectBest");

  graph.setFinishPoint("selectBest");

  return graph.compile();
}

File: server/graph/graph.py
from langgraph.graph import StateGraph, END
from typing import TypedDict, Optional, List, Any

from graph.nodes import (
    gpt_node, deepseek_node, nvidia_node, gemini_node,
    gpt_refine_node, deepseek_refine_node, nvidia_refine_node, gemini_refine_node,
    select_best_node,
)


class GraphState(TypedDict, total=False):
    question: str
    gpt: str
    deepseek: str
    nvidia: str
    gemini: str
    gpt_refined: str
    deepseek_refined: str
    nvidia_refined: str
    gemini_refined: str
    gpt_confidence: int
    deepseek_confidence: int
    nvidia_confidence: int
    gemini_confidence: int
    gpt_preferred: Optional[str]
    deepseek_preferred: Optional[str]
    nvidia_preferred: Optional[str]
    gemini_preferred: Optional[str]
    finalAnswer: str
    bestModel: str
    refinedAnswers: List[Any]
    peerReviews: List[Any]


def create_graph():
    g = StateGraph(GraphState)

    g.add_node("runGPT", gpt_node)
    g.add_node("runDeepSeek", deepseek_node)
    g.add_node("runNvidia", nvidia_node)
    g.add_node("runGemini", gemini_node)

    g.add_node("refineGPT", gpt_refine_node)
    g.add_node("refineDeepSeek", deepseek_refine_node)
    g.add_node("refineNvidia", nvidia_refine_node)
    g.add_node("refineGemini", gemini_refine_node)

    g.add_node("selectBest", select_best_node)

    g.set_entry_point("runGPT")

    g.add_edge("runGPT", "runDeepSeek")
    g.add_edge("runDeepSeek", "runNvidia")
    g.add_edge("runNvidia", "runGemini")

    g.add_edge("runGemini", "refineGPT")
    g.add_edge("refineGPT", "refineDeepSeek")
    g.add_edge("refineDeepSeek", "refineNvidia")
    g.add_edge("refineNvidia", "refineGemini")

    g.add_edge("refineGemini", "selectBest")
    g.add_edge("selectBest", END)

    return g.compile()
